src/Native/Unix/System.Native/pal_io.c

```diff
@@ -395,9 +394,6 @@ int32_t SystemNative_ReadDirR(DIR* dir, void* buffer, int32_t bufferSize, struct
         }
         return -1;
     }
-
-    assert(result->d_reclen <= bufferSize);
-    memcpy_s(entry, sizeof(struct dirent), result, (size_t)result->d_reclen);
```

Comment:
This changes makes it critical that GetDirectoryEntryName is called on the same thread as ReadDirR immediately after, right?  Previously ReadDirR could pass in its own memory and that memory would end up containing all of the data, but now it's relying on the data stored by readdir remaining valid through the call to GetDirectoryEntryName.  That should be fine, but it at least deserves a comment I think at the call site.

Reply:
> This changes makes it critical that GetDirectoryEntryName is called on the same thread as ReadDirR immediately after, right?

It's a bit more relaxed, the data is valid till the next call to readdir/closedir. But indeed, it should be copied as soon as possible.
I've updated the comments to reflect this more.